lms/templates/dashboard.html

```diff
@@ -108,28 +113,60 @@
+                entitlement = enrollment if isinstance(enrollment, CourseEntitlement) else None
+                entitlement_session = entitlement.enrollment_course_run if entitlement else None
+
+                is_fulfilled_entitlement = True if entitlement and entitlement_session else False
+                is_unfulfilled_entitlement = True if entitlement and not entitlement_session else False
+
+                entitlement_available_sessions = []
+                if entitlement:
+                  # Grab the available, enrollable sessions for a given entitlement and scrape them for relevant attributes
+                  entitlement_available_sessions = [{
+                    'session_id': course['key'],
+                    'enrollment_end': course['enrollment_end'],
+                    'pacing_type': course['pacing_type'],
+                    'session_start': CourseOverview.get_from_id(CourseKey.from_string(course['key'])).dashboard_start_display,
+                    'session_end': CourseOverview.get_from_id(CourseKey.from_string(course['key'])).dashboard_end_display,
+                    } for course in course_entitlement_available_sessions[str(entitlement.uuid)]]
+                  if is_fulfilled_entitlement:
+                    # If the user has a fulfilled entitlement, pass through the entitlements CourseEnrollment object
+                    enrollment = entitlement_session
+                  else:
+                    # If the user has an unfulfilled entitlement, pass through a bare CourseEnrollment object built off of the next available session
+                    upcoming_sessions = course_entitlement_available_sessions[str(entitlement.uuid)]
+                    next_session = upcoming_sessions[0] if upcoming_sessions else None
+                    if not next_session:
+                        continue
+                    enrollment = CourseEnrollment(user=user, course_id=next_session['key'], mode=next_session['type'])
+
+                session_id = enrollment.course_id
+                show_courseware_link = (session_id in show_courseware_links_for)
+                cert_status = cert_statuses.get(session_id)
+                can_unenroll = (not cert_status) or cert_status.get('can_unenroll') if not unfulfilled_entitlement else False
+                credit_status = credit_statuses.get(session_id)
+                show_email_settings = (session_id in show_email_settings_for)
+                course_mode_info = all_course_modes.get(session_id)
+                is_paid_course = True if entitlement else (session_id in enrolled_courses_either_paid)
+                is_course_blocked = (session_id in block_courses)
+                course_verification_status = verification_status_by_course.get(session_id, {})
+                course_requirements = courses_requirements_not_met.get(session_id)
+                related_programs = inverted_programs.get(unicode(session_id))
```

Comment:
Nit: we cast to `unicode` here, but on https://github.com/edx/edx-platform/pull/16472/files#diff-06f447d67e696470d5817797bfe9f7b9R148 we cast to `str`. We should try to use one type consistently throughout the code.

Reply:
Good point, I just copied this from a previous implementation but makes sense